File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {

public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *sentinal = new ListNode(INT_MIN, head);
        ListNode *leader = head;
        ListNode *cur = head;
        ListNode *prev = sentinal;
        
        for( int i = 0; i < n; i++)
            leader = leader->next;
        while( leader ) {
            leader = leader->next;
            prev = cur;
            cur = cur->next;
        }
        
        prev->next = cur->next;
        head = sentinal->next;
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        
        sentinel = ListNode(float("-inf"), head)
        
        leader = head
        for _ in range(n):
            leader = leader.next
        cur = head
        prev = sentinel
  
        while leader is not None:
            leader = leader.next
            prev = cur
            cur = cur.next
        
        prev.next = cur.next
        
        return sentinel.next
        